toolkit/tools/internal/grub/grubtokenizer.go

```diff
@@ -0,0 +1,509 @@
+	}
+
+	return tokenizer.tokens, nil
+}
+
+type grubConfigTokenizer struct {
+	scanner  *filescanner.FileScanner
+	tokens   []Token
+	subWords []SubWord
+}
+
+func (t *grubConfigTokenizer) tokenize() error {
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break
+		}
+
+		switch c {
+		// Metacharacters
+		case '{', '}', '|', '&', ';', '<', '>', '\n':
+			locStart := t.scanner.Loc()
+			// Consume metacharacter
+			t.scanner.Next()
+
+			var tokenType TokenType
+			switch c {
+			case '{':
+				tokenType = LBRACE
+			case '}':
+				tokenType = RBRACE
+			case '|':
+				tokenType = BAR
+			case '&':
+				tokenType = AND
+			case ';':
+				tokenType = SEMICOLON
+			case '<':
+				tokenType = LT
+			case '>':
+				tokenType = GT
+			case '\n':
+				tokenType = NEWLINE
+			}
+
+			token := t.newToken(locStart, t.scanner.Loc(), tokenType)
+			t.tokens = append(t.tokens, token)
+
+		// Space
+		case ' ', '\t':
+			err := t.parseSpace()
+			if err != nil {
+				return err
+			}
+
+		default:
+			err := t.parseWord()
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func (t *grubConfigTokenizer) parseWord() error {
+	locStart := t.scanner.Loc()
+
+	c, eof := t.scanner.Peek()
+	if eof {
+		return nil
+	}
+
+	// Check if the word starts with '#'.
+	// Note: A '#' in the middle of a word doesn't start a comment.
+	if c == '#' {
+		err := t.parseComment()
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	var err error
+	t.subWords = nil
+	notFirst := false
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break
+		}
+
+		switch c {
+		// Metacharacters and spaces
+		case '{', '}', '|', '&', ';', '<', '>', '\n', ' ', '\t':
+			break loop
+
+		// Double-quoted string
+		case '"':
+			err = t.parseDoubleQuotedString()
+			if err != nil {
+				break loop
+			}
+
+		// Single-quoted string
+		case '\'':
+			err = t.parseSingleQuotedString()
+			if err != nil {
+				break loop
+			}
+
+		// Variable expansion
+		case '$':
+			err = t.parseVariableExpansion(VAR_EXPANSION)
+			if err != nil {
+				break loop
+			}
+
+		// Unquoted string
+		default:
+			err = t.parseUnquotedString(notFirst)
+			if err != nil {
+				break loop
+			}
+		}
+
+		notFirst = true
+	}
+
+	token := t.newToken(locStart, t.scanner.Loc(), WORD)
+	token.SubWords = t.subWords
+	t.subWords = nil
+	t.tokens = append(t.tokens, token)
+	return err
+}
+
+func (t *grubConfigTokenizer) parseSpace() error {
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break loop
+		}
+
+		switch c {
+		case ' ', '\t':
+			t.scanner.Next()
+
+		default:
+			break loop
+		}
+	}
+
+	return nil
+}
+
+func (t *grubConfigTokenizer) parseComment() error {
+	// Consume the '#' char.
+	t.scanner.Next()
+
+	sb := strings.Builder{}
+loop:
+	for {
+		c, eof := t.scanner.Peek()
+		if eof {
+			break loop
+		}
+
+		switch c {
+		case '\n':
+			break loop
+
+		default:
+			sb.WriteRune(c)
```

Comment:
sb (the comment) is not captured, correct?

Reply:
should the parser be able to re-construct the input file from the tokens? if so, will the commends be preserved?

Reply:
Old code. I originally gave comments their own token type. But decided that made it more difficult to interpret the tokens.

As it stands, you can't perfectly reconstruct the original file due to missing tokens for whitespace and comments.